api/bindings.h

```diff
@@ -88,13 +100,27 @@ typedef struct db_t {
   uint8_t _private[0];
 } db_t;
 
+/**
+ * A view into a `Option<&[u8]>`, created and maintained by Rust.
+ *
+ * This can be copied into a []byte in Go.
+ */
+typedef struct U8SliceView {
+  /**
+   * True if and only if this is None/nil. If this is true, the other fields must be ignored.
+   */
+  bool is_nil;
+  const uint8_t *ptr;
+  uintptr_t len;
+} U8SliceView;
+
 typedef struct iterator_t {
   uint64_t db_counter;
   uint64_t iterator_index;
 } iterator_t;
 
 typedef struct Iterator_vtable {
-  int32_t (*next_db)(iterator_t, gas_meter_t*, uint64_t*, Buffer*, Buffer*, Buffer*);
+  int32_t (*next_db)(iterator_t, gas_meter_t*, uint64_t*, UnmanagedVector*, UnmanagedVector*, UnmanagedVector*);
```

Comment:
I'm confused.
Those fields must be written by Go.

Reply:
Right, but Go does not write it directly. It instructs Rust to create a now UnmanagedVector and overrides the pointer with it.

Any UnmanagedVector that was at that pointer before is lost. This is why we ensure they do not have a heap allocation which is leaked in:

```go
	if !(*key).is_nil || !(*val).is_nil || !(*errOut).is_nil {
		panic("Got a non-nil UnmanagedVector we're about to override. This is a bug because someone has to drop the old one.")
	}
```

Reply:
This is double indirection because we use pointers for return values.

A pointer into memory holding the `UnmanagedVector` struct, which itself is an optional pointer to some heap data.